WIP: await promise code from sync c

hedwigz opened the pull request:
Introduced a new macro `MAIN_THREAD_EM_ASM_PROMISE_AWAIT` which is used to write javascript code that **returns a promise** and block the C code from progressing until the promise is resolved (or errors).  
```C
int main()
{
  printf("1st print\n");
  // This call will block until the internal promise resolves (at least 1 second in this case) 
  MAIN_THREAD_EM_ASM_PROMISE_AWAIT({
    out('2nd print');
    // do some async operation that involves the main thread
    return new Promise((resolve,reject) => {
      setTimeout(() => {
        out('3rd print');
        resolve();
      }, 1000);
    });
  });
  printf("4th print\n");
  return 0;
}
```
The motivation for this was [discussed before](https://github.com/emscripten-core/emscripten/discussions/21929)  
We have multiple web workers, each of them calls javascript code which is async and has to be ran on the main thread and need to wait for the promise to be resolved.  
I am open for suggestions on how to add this more elegantly

TODO:

- [x] add test for promise rejection
- [x] handle call from main thread
- [x] support returning value (or acknowledging known-issue)
- [x] docs

hedwigz commented on src/library_pthread.js:
```diff
@@ -983,6 +983,27 @@ var LibraryPThread = {
+#endif
+      __emscripten_proxy_promise_finish(res, promiseCtx);
+      }).catch(err => {
+        __emscripten_proxy_promise_finish(err, promiseCtx);
```
It's defined [here](https://github.com/emscripten-core/emscripten/pull/23043/files#diff-ca572ec5c79d581d3657cfcf00705c200889335d5289c917656864187fb6ca62R651-R653)  
I was not sure how to add an internal C function that is available to the runtime so I added it to `REQUIRED_EXPORTS` [here](https://github.com/emscripten-core/emscripten/pull/23043/files#diff-a9cba4e03450a831038c69937bac7910c60b2a341f1eef13b70aaf4e21cb7c5dR505)  
Is there a better way to do this?

hedwigz commented on test/core/test_main_thread_async_em_asm_promise_await.cpp:
```diff
@@ -0,0 +1,24 @@
+// Copyright 2017 The Emscripten Authors.  All rights reserved.
```
fixed

hedwigz commented on test/test_core.py:
```diff
@@ -1896,6 +1896,13 @@ def test_main_thread_em_asm(self, args):
 
+  @needs_dylink
+  @parameterized({
+    'pthreads': (['-pthread', '-sPROXY_TO_PTHREAD', '-sEXIT_RUNTIME'], False),
```
I added the empty one, and the test expects for an error. I did not handle the case of calling this macro from the main thread. Do you think we should add that?

hedwigz commented:
@sbc100 ping 🤞

sbc100 commented on src/library.js:
```diff
@@ -1655,6 +1655,16 @@ addToLibrary({
   emscripten_asm_const_int_sync_on_main_thread: (emAsmAddr, sigPtr, argbuf) => runMainThreadEmAsm(emAsmAddr, sigPtr, argbuf, 1),
 
+  emscripten_asm_const_int_await_promise_on_main_thread__deps: ['$runMainThreadEmAsm'],
+  emscripten_asm_const_int_await_promise_on_main_thread: (emAsmAddr, sigPtr, argbuf) => {
```
How about we drop the `promise` part so this becomes `emscripten_asm_const_int_await_on_main_thread` and the macro is `MAIN_THREAD_EM_ASM_AWAIT`?

hedwigz replied:
Yes, I like that better as well

sbc100 commented on src/library.js:
```diff
@@ -1631,7 +1631,7 @@ addToLibrary({
 #endif
   ],
-  $runMainThreadEmAsm: (emAsmAddr, sigPtr, argbuf, sync) => {
+  $runMainThreadEmAsm: (emAsmAddr, sigPtr, argbuf, sync, promise) => {
```
Maybe call this argument "await" too.. although that could be confused with the JS keyword.  Maybe "asyncAwait"?

hedwigz replied:
👍

sbc100 commented on src/library.js:
```diff
@@ -1655,6 +1655,16 @@ addToLibrary({
+      return runMainThreadEmAsm(emAsmAddr, sigPtr, argbuf, 1, 1);
+    }
+    #endif
+    throw new Error('call to emscripten_asm_const_int_await_promise_on_main_thread is only supported from pthread (but was called from main thread)');
```
We probably don't want this long error message in release builds.  How about:

```
#if ASSERTIONS
assert(ENVIRONMENT_IS_PTHREAD. "emscripten_asm_const_int_await_promise_on_main_thread is not available on the main thread");
#endif
return runMainThreadEmAsm(emAsmAddr, sigPtr, argbuf, 1, 1);
```

hedwigz replied:
👍

hedwigz replied:
I had to modify the assert condition because it errored: `ReferenceError: ENVIRONMENT_IS_PTHREAD is not defined`  
```javascript
assert((typeof ENVIRONMENT_IS_PTHREAD !== 'undefined' && ENVIRONMENT_IS_PTHREAD), "emscripten_asm_const_int_await_on_main_thread is not available on the main thread");
```
Not sure if there's a prettier way

sbc100 replied:
Use `#if ASSERTION && PTHREADS`

sbc100 commented on src/library.js:
```diff
@@ -1655,6 +1655,16 @@ addToLibrary({
+  emscripten_asm_const_int_await_promise_on_main_thread: (emAsmAddr, sigPtr, argbuf) => {
+    #if PTHREADS
+    if (ENVIRONMENT_IS_PTHREAD) {
+      return runMainThreadEmAsm(emAsmAddr, sigPtr, argbuf, 1, 1);
```
Can you add comments here for these contants e.g.

```
return runMainThreadEmAsm(emAsmAddr, sigPtr, argbuf, /*sync=*/1, /*asyncAwait=*/1);
```

hedwigz replied:
👍

sbc100 commented on src/library_pthread.js:
```diff
@@ -983,6 +983,27 @@ var LibraryPThread = {
+#endif
+      if (!rtn.then) {
+        throw new Error('Return value of proxied function expected to be a Promise but got' + rtn);
+      }
```
Need for both that assertion and runtime check here.  I think you can just remove these 3 lines.

hedwigz replied:
oops. nice catch

sbc100 commented on system/lib/pthread/proxying.c:
```diff
@@ -583,31 +632,35 @@ em_promise_t emscripten_proxy_promise(em_proxying_queue* q,
+    f->funcIndex, f->emAsmAddr, f->callingThread, f->numArgs, f->argBuffer, (void*)f->ctx);
+}
+
+void _emscripten_proxy_promise_finish(void* res, em_proxying_ctx* ctx) {
```
I would expect the context to be the first arugment here.

hedwigz replied:
👍

sbc100 commented on system/lib/pthread/proxying.c:
```diff
@@ -620,12 +673,18 @@ double _emscripten_run_on_main_thread_js(int func_index,
+      return 0;
+    }
+    return f.result;
   }
```
Maybe structure this as:

```
if (sync) {
   if (promise) {
   } else { 
   }
   return f.result;
}
```

Also the fact that we have `int sync` and `int promise` here as a separate arguments, but it doesn't make sense for `promise` to be set without `sync` seem a little odd.

At the very least we should probably assert that promise is never true when sync is not.

tlively commented on system/lib/pthread/proxying.c:
Since the new code in this file does not implement the API surface in proxying.h, it would be best to move it to a different file and leave proxying.c unchanged. It would also be best to implement the new functionality in terms of the public API in proxying.h rather than using internal implementation details. For example, there should be no need to reach inside the `em_proxying_ctx` object.

tlively replied:
Although I see we already have `_emscripten_run_on_main_thread_js` at the end of proxying.c. Adding more code there would be fine, but it should still use only the public API from proxying.h.

hedwigz replied:
Do you prefer that I introduce a new function similar to `_emscripten_run_on_main_thread_js` but that uses only public proxying.c API?
The problem with introducing a new function similar to `emscripten_run_on_main_thread_js` is the "path" to that function call.  
It'd a bit tedious. And it'd modify the whole stack leading to this logic.  
This is why I chose to introduce the new function param `asyncAwait` (previously `promise`) just so I could "ride" the whole functions call path.  
But, I don't insist on this at all.  
As for using the public API - I agree that we can do that if I indeed fork from `_emscripten_run_on_main_thread_js`  
Let me know what you think

tlively replied:
Why do you need to use internal details of the proxying mechanism if you don't create an entirely new version of `_emscripten_run_on_main_thread_js`? Can't you use `emscripten_proxy_sync_with_ctx` to synchronously wait for the promise to be resolved on the main thread instead of introducing a new `emscripten_proxy_async_await` function?

hedwigz replied:
The challenging part was to pass the `em_proxying_ctx` all the way down to `_emscripten_receive_on_main_thread_js`, through the `task` and `proxied_js_func_t` wrappers, because I have to call `_emscripten_proxy_promise_finish` from within `_emscripten_receive_on_main_thread_js`.  
This is also why I added `em_proxying_ctx * ctx` to `proxied_js_func_t` and moved it upwards.  
I made a significant simplification and deleted most of the previous code in `proxying.c`  
WDYT?

hedwigz replied:
Now that I think of it, it is a little weird that `proxied_js_func_t` has a field `ctx` that if set, causes `_emscripten_receive_on_main_thread_js` to expect a promise. Maybe I should name it differently?

tlively replied:
I think the `ctx` and `promiseCtx` names are fine, but it would be good to add comments explaining that they can be null.

hedwigz commented:
@sbc100 @tlively 🙏 I really appreciate the thorough review and comments!  
I addressed all of your comments and I think the code looks way better now.

tlively commented on system/lib/pthread/proxying.c:
```diff
@@ -642,3 +651,34 @@ double _emscripten_run_on_main_thread_js(int func_index,
   return 0;
 }
+
+double _emscripten_await_on_main_thread_js(int func_index,
```
I do think there is enough code shared with `_emscripten_run_on_main_thread_js` that it may make sense to combine them, but I don't feel strongly about it. If you do keep them separate, you should explicitly initialize the `.ctx` field to `NULL` in `_emscripten_run_on_main_thread_js` to make sure it is initialized .

tlively commented on system/lib/pthread/proxying.c:
```diff
@@ -642,3 +651,34 @@ double _emscripten_run_on_main_thread_js(int func_index,
+  proxied_js_func_t* func = (proxied_js_func_t*)t->arg;
+  func->result = (double)(intptr_t)res;
+  emscripten_proxy_finish(ctx);
+}
```
Please make sure there is a newline at the end of the file. It would also be good to run `git clang-format` to fix any formatting errors.

tlively reviewed (commented):
Great, this looks a lot simpler overall, even though I agree that it's a little awkward to get the `ctx` into the right place. Thanks!

hedwigz commented:
> Great, this looks a lot simpler overall, even though I agree that it's a little awkward to get the `ctx` into the right place. Thanks!

Awesome :)  
Any tips for how to pass the CI tests?

tlively commented:
proxying.c looks good to me.

It looks like the flake8 bot is failing because of a formatting issue. The other CI bots might be aborting early because of that issue. Merging in main to pick up any unrelated fixes could also help.

sbc100 commented:
The fact that flake8 is being included at all in CI makes me think that perhaps you are not up-to-date on your branch?  The flake8 step was replaced with a step call `ruff` few days back.

sbc100 commented:
Can you try rebase onto to merging with the latest changes on main?

hybridherbst commented:
Any updates or blockers here? I'd like to see this land as well, we're also using workers that need to sync await some main thread code.

hedwigz commented:
> Any updates or blockers here? I'd like to see this land as well, we're also using workers that need to sync await some main thread code.

Hey @hybridherbst , glad to see more interest in this.  The current blocker is that some wasm64 test fail. I was struggling making the test run on my machine due to the need in d8/node24

hedwigz commented:
@sbc100 Thanks! I'll give it a go

hedwigz commented:
@tlively @sbc100 All tests are now green :)

sbc100 commented on test/test_core.py:
```diff
@@ -1892,6 +1892,26 @@ def test_main_thread_em_asm(self, args):
+  def test_main_thread_async_em_asm_await(self, args, force_c=False):
+    if '-sPROXY_TO_PTHREAD' not in args:
+      # expect runtime to error
+      output = self.do_runf('core/test_main_thread_async_em_asm_await.cpp', expected_output=None, assert_returncode=NON_ZERO, emcc_args=args)
```
You can remove `expected_output=None` here

sbc100 commented on test/test_core.py:
```diff
@@ -1892,6 +1892,26 @@ def test_main_thread_em_asm(self, args):
+    '': (['-sASSERTIONS'], False),
+    'pthreads': (['-pthread', '-sPROXY_TO_PTHREAD', '-sEXIT_RUNTIME'], False),
+  })
+  def test_main_thread_async_em_asm_await(self, args, force_c=False):
```
It looks like force_c is always false.  Maybe just remove it?

sbc100 commented on test/core/test_main_thread_async_em_asm_await_reject.cpp:
```diff
@@ -0,0 +1,34 @@
+#include <pthread.h>
+
+int main()
+{
```
Curly brace goes on the same line as the function name

sbc100 commented on test/core/test_main_thread_async_em_asm_await_reject.cpp:
```diff
@@ -0,0 +1,34 @@
+// Copyright 2024 The Emscripten Authors.  All rights reserved.
```
2025

sbc100 commented on test/core/test_main_thread_async_em_asm_await_reject.cpp:
```diff
@@ -0,0 +1,34 @@
+{
+  // start new thread
+  pthread_t thread;
+  pthread_create(&thread, NULL, [](void*) -> void* {
```
Can you split this out into a separate function

sbc100 commented on tools/link.py:
```diff
@@ -514,6 +514,7 @@ def setup_pthreads():
   settings.REQUIRED_EXPORTS += [
     '_emscripten_thread_free_data',
     '_emscripten_thread_crashed',
+    '_emscripten_proxy_promise_finish',
```
I don't think this should be needed here.  At least I don't see runtime_pthread.js depending on this.  Instead I think you can add this to `_emscripten_receive_on_main_thread_js__deps`

sbc100 commented on src/lib/libpthread.js:
```diff
@@ -960,6 +970,24 @@ var LibraryPThread = {
     PThread.currentProxiedOperationCallerThread = 0;
+    if (promiseCtx) {
+#if ASSERTIONS
+      assert(!!rtn.then, 'Return value of proxied function expected to be a Promise but got' + rtn);
```
I don't think you need the `!!` here ?

sbc100 commented on src/lib/libpthread.js:
```diff
@@ -960,6 +970,24 @@ var LibraryPThread = {
+#if ASSERTIONS
+      assert(!!rtn.then, 'Return value of proxied function expected to be a Promise but got' + rtn);
+#endif
+      rtn.then(res => {
```
We put braces around the `res` here and `err` below.

sbc100 commented on system/lib/pthread/proxying.c:
```diff
@@ -591,12 +591,15 @@ typedef struct proxied_js_func_t {
   bool owned;
+  // Only used when the underlying js func is async.
+  // Can be null when the function is sync.
+  em_proxying_ctx * ctx;
```
No space before `*` here.

sbc100 commented on test/test_core.py:
```diff
@@ -1972,6 +1972,24 @@ def test_main_thread_em_asm(self, args):
+      output = self.do_runf('core/test_main_thread_async_em_asm_await.cpp', assert_returncode=NON_ZERO, emcc_args=args)
+      self.assertContained('emscripten_asm_const_int_await_on_main_thread is not available on the main thread', output)
+    else:
+      self.do_core_test('test_main_thread_async_em_asm_await.cpp', emcc_args=args, force_c=False)
```
You don't need `force_c=False` here

sbc100 commented on test/test_core.py:
```diff
@@ -1972,6 +1972,24 @@ def test_main_thread_em_asm(self, args):
+    'pthreads': (['-pthread', '-sPROXY_TO_PTHREAD', '-sEXIT_RUNTIME', '-sASSERTIONS'], False),
+  })
+  def test_main_thread_async_em_asm_await_reject(self, args, force_c=False):
+    self.do_core_test('test_main_thread_async_em_asm_await_reject.cpp', emcc_args=args, force_c=force_c)
```
I don't think you need `force_c` here or above.

sbc100 commented on test/test_core.py:
```diff
@@ -1972,6 +1972,24 @@ def test_main_thread_em_asm(self, args):
+
+  @parameterized({
+    'pthreads': (['-pthread', '-sPROXY_TO_PTHREAD', '-sEXIT_RUNTIME', '-sASSERTIONS'], False),
+  })
```
I think you can just remove the `@parameterized` block since there is only one variant of this test.

sbc100 commented on test/core/test_main_thread_async_em_asm_await.cpp:
```diff
@@ -0,0 +1,25 @@
+#include <stdio.h>
+
+int main()
+{
```
Open brace on same line as function decl.

sbc100 commented on test/core/test_main_thread_async_em_asm_await_reject.cpp:
```diff
@@ -0,0 +1,35 @@
+    return asyncOp;
+  }, 42, 3.5);
+  printf("After MAIN_THREAD_EM_ASM_AWAIT rejected\n");
+  printf("result: %d\n", res);
```
If code that uses threads its better to use `emscripten_outf` rather than `printf` since its writes to the lower level file descriptor under node. (Writing to stdout can have buffering issues).

sbc100 commented:
Thanks for continuing to work on this.

Apologies for the review process taking so long.  I had no imagined this being such an invasive change (i.e.changing the code proxying functions).
